Added script that checks links in a file.

Delet0r opened the pull request:
All the links in src/data.js are hardcoded. This introduces the
possiblity of them breaking without anyone noticing. To be able
to automate the checking of all links this script is introduced.

The only parameter this script takes is a file path. The file is then
read and parsed for all occurences of http/https links. Those
links are then checked if they return an http status of 2** if not
the failing links are printed to the command line with their
returned status code.

Commit:
Added script that checks links in a file.

All the links in src/data.js are hardcoded. This introduces the
possiblity of them breaking without anyone noticing. To be able
to automate the checking of all links this script is introduced.

The only parameter this script takes is a file path. The file is then
read and parsed for all occurences of http/https links. Those
links are then checked if they return an http status of 2** if not
the failing links are printed to the command line with their
returned status code.

Commit:
Cleaned up imports by making them more specific

cpetig commented on helper_scripts/check_links/check_links.py:
```diff
@@ -0,0 +1,41 @@
+
+            erroneous_responses = filter(lambda response: str(response.code)[0] != '2', responses)
+
+            print("The following links in `data.js` are not returning success status codes (2**):")
```
I believe "data.js" should be  "%{args.file_path}" (not sure about the correct syntax here)

Delet0r replied:
Ah thanks, I totally overlooked that.

Delet0r replied:
Fixed

cpetig commented on helper_scripts/check_links/check_links.py:
```diff
@@ -0,0 +1,41 @@
+            for  response in erroneous_responses:
+                print("    {0}: {1}".format(response.url, response.code))
+    except FileNotFoundError as e:
+        print("Error: The given path does not point to a valid file.")
```
It might be obvious here but I tend to repeat/state the path name in error messages complaining about missing files. If you wrap the script in some framework it might easily get less obvious.

Delet0r replied:
Fixed

cpetig reviewed (commented):
The script basically does what it is supposed to do, but after receiving more than one screen full of errors about weak encryption on https://www.saarland.de/finanzaemter.htm I think the output might easily confuse less tech savvy users. (And I don't know whether this error prevents other URLS from being checked or not)

This happens on ubuntu 20.04 with python3:
$ python3 ../../check_links.py data.js 
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 485, in wrap_socket
    cnx.do_handshake()
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1915, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
    _raise_current_error()
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'ssl_choose_client_version', 'unsupported protocol')]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 376, in _make_request
    self._validate_conn(conn)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 996, in _validate_conn
    conn.connect()
  File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 352, in connect
    self.sock = ssl_wrap_socket(
  File "/usr/lib/python3/dist-packages/urllib3/util/ssl_.py", line 370, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 491, in wrap_socket
    raise ssl.SSLError("bad handshake: %r" % e)
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'ssl_choose_client_version', 'unsupported protocol')])",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 719, in urlopen
    retries = retries.increment(
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 436, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='www.saarland.de', port=443): Max retries exceeded with url: /finanzaemter.htm (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl_choose_client_version', 'unsupported protocol')])")))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "../../check_links.py", line 39, in <module>
    main()
  File "../../check_links.py", line 27, in main
    responses = [Response(url, get(url, timeout=10).status_code) for url in urls]
  File "../../check_links.py", line 27, in <listcomp>
    responses = [Response(url, get(url, timeout=10).status_code) for url in urls]
  File "/usr/lib/python3/dist-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='www.saarland.de', port=443): Max retries exceeded with url: /finanzaemter.htm (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl_choose_client_version', 'unsupported protocol')])")))

Delet0r commented:
Hm, I can't reproduce the error. Could you give me the version of your `requests` library?

I could filter the status code errors to 4** only and also display what the error means. Do you think that this is sufficient to prevent confusion with non tech savvy users?

Commit:
Fixed request from PR review

- replaced hardcoded filename with file path variable
- added file path to error message to be more informativ

cpetig commented:
python3-urllib3                      1.25.8-2
python3-openssl                      19.0.0-1build1 
I guess this is an error generated from the openSSL settings - I will dig deeper into it.

cpetig commented:
The site uses SSL1.0 (and rejecting this outdated protocol is quite common)

Commit:
Fixed request from PR review round 2

Disabled SSL verification for the `urlopen` call. This was necesarry,
since it created a false negativ, where an url was listed as broken
even though the server "just" used no current SSL version.

Rephrased parts of the scripts output to be more understandable for non
techy users (hopefully :)) and added the phrasing of the HTTP Errors to
the output.

Moved the except clause of the try statement (for opening the file)
closer together. The code in between was unnecessarly part of the try
statement because it had nothing to do with the opened file descriptor.

Delet0r commented:
With the last commit SSL verification was disabled and the false negatives therefor fixed. Unfortunately this didn't fix the problem of @cpetig. As of now I was not able to reproduce the error. Until we get more cases of this error it will be regarded as unimportant.

Commit:
Made script mostly pylint conform.
